### PR TITLE
Enforcement

### PR DESCRIPTION
**Pull request by atz**
Changes appear larger than they actually are, but include:
- Correct comment about solr_access_filters_logic params
- Un-protect `discovery_permissions` getter/setter
- Use attribute and alias declarations where possible
- Lots more YARD doc for all
- Delete `#except` and `#append` tests, these were testing SearchBuilder, not Enforcement.  Apparently the methods existed in Enforcement once, but were deleted.  The tests remained and still passed, so that gives you an idea of how useful they were.
- Stop testing SearchBuilder, start testing a class including the module.  That class represents our downstream consumers, not SearchBuilder.
- Additional tests for a couple basic operations and exclusion of empty clauses

### CODE REVIEW

File: lib/blacklight/access_controls/enforcement.rb [hunk 1]
```diff
@@ -25,72 +29,61 @@ module Enforcement
...
 
+      # For groups
+      # @return [Array{String}] values are lucence syntax term queries suitable for :fq
+      # @example
+      #   [ "({!terms f=discover_access_group_ssim}public,faculty,africana-faculty,registered)",
+      #     "({!terms f=read_access_group_ssim}public,faculty,africana-faculty,registered)" ]
       def apply_group_permissions(permission_types, ability = current_ability)
-        # for groups
+        groups = ability.user_groups
+        return [] if groups.empty?
```
Comment (jcoyne) [inline lib/blacklight/access_controls/enforcement.rb]: doesn't thin now apply no permissions, which gives access to everything?

File: spec/unit/enforcement_spec.rb [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
+  include Blacklight::AccessControls::Enforcement
+end
+
 describe Blacklight::AccessControls::Enforcement do
-  let(:controller) { CatalogController.new }
-  let(:search_builder) { SearchBuilder.new(method_chain, controller) }
-  let(:method_chain) { SearchBuilder.default_processor_chain }
+  let(:controller) do
+    c = MyController.new
+    allow(c).to receive(:current_ability).and_return(ability)
```
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: Why are you stubing a method on a class you defined? why not just override `current_ability` in the class definition.

File: spec/unit/enforcement_spec.rb [hunk 3]
```diff
@@ -1,15 +1,29 @@
 # frozen_string_literal: true
+class MyController # < ApplicationController
```
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: This class should probably be destroyed after running the tests, otherwise you could have collisions with other tests.

File: spec/unit/enforcement_spec.rb [hunk 4]
```diff
@@ -19,10 +33,20 @@
       solr_parameters[:fq].first
     end
 
+    # rubocop:disable RSpec/MessageExpectation
+    describe 'logger' do
+      let(:ability) { instance_double(Ability, user_groups: [], current_user: user) } # prevent Ability from hitting logger first
```
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: The comment is unclear to me. Can you provide more explanation?

File: spec/unit/enforcement_spec.rb [hunk 5]
```diff
@@ -19,10 +33,20 @@
       solr_parameters[:fq].first
     end
 
+    # rubocop:disable RSpec/MessageExpectation
+    describe 'logger' do
```
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: is this testing the logger method, or is it testing that `apply_gated_discovery` is calling the logger?

File: spec/unit/enforcement_spec.rb [hunk 6]
```diff
@@ -46,6 +70,11 @@
         expect(fq_first).to match(%r{\{!terms f=read_access_group_ssim\}public,faculty,africana-faculty,registered})
       end
 
+      it 'does not build empty clauses' do
+        expect(controller).to receive(:apply_user_permissions).and_return(['({!terms f=discover_access_group_ssim}public,faculty,africana-faculty,registered)', '', nil])
+        expect(fq_first).not_to match(/ OR $/) # i.e. doesn't end w/ emtpy
```
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: typo "empty"

File: lib/blacklight/access_controls/enforcement.rb [hunk 7]
```diff
@@ -25,72 +29,61 @@ module Enforcement
...
+        return [] unless user && user.user_key.present?
+        permission_types.map do |type|
+          escape_filter(solr_field_for(type, 'user'), user.user_key)
         end
-        user_access_filters
       end
 
-      # Find the name of the solr field for this type of permission.
-      # e.g. "read_access_group_ssim" or "discover_access_person_ssim".
+      # @param [#to_s] permission_type, a single value, e.g. "read" or "discover"
```
Comment (jcoyne) [inline lib/blacklight/access_controls/enforcement.rb]: Don't use a comma after the param name see http://www.rubydoc.info/gems/yard/file/docs/Tags.md#param

File: spec/unit/enforcement_spec.rb (hunk 4 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: The expectation set will be falsely triggered by Ability class (that also calls Rails.logger.debug frequently).  So we use a double'd version of Ability to avoid tripping that code.

File: spec/unit/enforcement_spec.rb (hunk 5 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: The latter.

File: lib/blacklight/access_controls/enforcement.rb (hunk 1 above)
Comment (atz) [inline lib/blacklight/access_controls/enforcement.rb]: Seeing as how Ability has default groups and even anonymous public users have a group, I might also accept throwing an error here.  But if you do intentionally pass "no group filters", what do you expect?  Multiple `:fq`'s against nothing or no group filters?  

If you don't have groups then you would be creating filter clauses of, e.g. `({!terms f=my_field_ssim})`.  That matches nothing in a lucene query, because lucene doesn't index nothing.  But more broadly, the end result is an OR query.  After a single populated query, including additional no-match clauses is worthless.  Without groups, the compound `fq` is going to say, e.g.: 
`'discover_access_person_ssim:user_1@abc.com' OR 'read_access_person_ssim:user_1@abc.com'`

So that cannot match *more* than adding a bunch of other OR clauses.

Concrete queries:
[{!terms%20f=format_main_ssim}](https://sul-solr-a.stanford.edu/solr/exhibits_stage/select?fq={!terms%20f=format_main_ssim}&indent=on&wt=json) -- field exists, query nothing, get nothing
[{!terms%20f=format_main_ssim}Book](https://sul-solr-a.stanford.edu/solr/exhibits_stage/select?fq={!terms%20f=format_main_ssim}Book&indent=on&wt=json) -- field exists, query value, gets results
[{!terms f=no_such_field}](https://sul-solr-a.stanford.edu:443/solr/exhibits_stage/select?fq={!terms f=no_such_field}&indent=on&wt=json) -- field non-existent, query nothing, get nothing

File: spec/unit/enforcement_spec.rb (hunk 2 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: Was adapted from the old line 12 predating the class.  Can't invoke the let statements outside the describe block though, so this is still does the job.

File: spec/unit/enforcement_spec.rb (hunk 3 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: Are you actually worried about namespace collision with another `MyController`?  From where?  I don't get the concern as a practical matter.  If I move the class into `spec/support/` or `spec/helpers/` what has changed?  And yet defining test classes there is commonplace enough.  CurationConcerns defines classes in spec **19** times.  I think we can handle this tiny one here.

**Comment by atz** [comment]
@jcoyne Corrected the typos and added replies to your comments.

File: spec/unit/enforcement_spec.rb (hunk 4 above)
Comment (jcoyne) [inline spec/unit/enforcement_spec.rb]: Can you put that in the comment?

Comment (atz) [inline spec/unit/enforcement_spec.rb]: Sure.

**Comment by atz** [comment]
@jcoyne updated that comment

File: lib/blacklight/access_controls/enforcement.rb (hunk 1 above)
Comment (atz) [inline lib/blacklight/access_controls/enforcement.rb]: @jcoyne Do you consider this incorrect?  If so, then I would consider the parallel case to `apply_group_permissions` in `apply_user_permissions` also incorrect:
```ruby
return [] unless user && user.user_key.present?
```
While I cleaned up the form here, that behavior was pre-existing.  No user_key, no filters.  That was the scenario that actually had tests backing it, so I regarded it as more intentional:
https://github.com/projectblacklight/blacklight-access_controls/blob/eeac826692c056d55db8ef94a9a12cfa08578a92/spec/unit/enforcement_spec.rb#L134-L147

Happy to adapt if there is something more formal to work from, but this is my best understanding of the intended behavior.

Comment (jcoyne) [inline lib/blacklight/access_controls/enforcement.rb]: I don't know if it is or isn't. I'm just checking to see if you've considered everything. I'm not eager to make changes to something that we believe works okay and is fundamental to access enforcement.

Comment (atz) [inline lib/blacklight/access_controls/enforcement.rb]: The belief in this case is not validated by a test case or piece of documentation.  Also, pragmatically, since even anonymous users should have a group, in theory this case is excluded in well behaved systems.  I don't know what else to say.  If you don't know what is intended, there's only one other person on this file's history to ask.  @val99erie, do you have any insight here?

File: lib/blacklight/access_controls/enforcement.rb [hunk 8]
```diff
@@ -25,72 +29,61 @@ module Enforcement
...
-      # * Applies a lucene query to the solr :q parameter for gated discovery
-      # * Uses public_qt search handler if user does not have "read" permissions
-      # @param solr_parameters the current solr parameters
-      def add_access_controls_to_solr_params(solr_parameters)
-        apply_gated_discovery(solr_parameters)
-      end
-
       # Which permission levels (logical OR) will grant you the ability to discover documents in a search.
-      # Override this method if you want it to be something other than the default
+      # Override this method if you want it to be something other than the default, or hit the setter
```
Comment (cam156) [inline lib/blacklight/access_controls/enforcement.rb]: @atz If the new standard is to set the discovery_permissions via discovery_permissions=  The comment might be better to say that "This method defaults to [discover, read] permissions unless they have been set by calling  discovery_permissions=".  I think hit the setter may not entirely make sense to everyone.

File: lib/blacklight/access_controls/enforcement.rb (hunk 1 above)
Comment (cam156) [inline lib/blacklight/access_controls/enforcement.rb]: @atz have you looked at the sorl query created before and after your change?  Can you give an example?  We may be able to eyeball a test.  aka if the output is invalid or does not make any sense your code would be the better option.

File: spec/unit/enforcement_spec.rb (hunk 3 above)
Comment (cam156) [inline spec/unit/enforcement_spec.rb]: @atz The general practice has been for each spec to clean up after itself so that other specs run the same way if this test was run with it or not.  I believe all @jcoyne is asking for is for you to do is `Object.send(:remove_const, :MyController)` after the test

File: spec/unit/enforcement_spec.rb (hunk 2 above)
Comment (cam156) [inline spec/unit/enforcement_spec.rb]: @atz you can define the class within the describe block, and then include a method that returns ability.  Then use additional let statements to set that value. @jcoyne do you have an example?  I spent a little bit of time looking through sufia and cc and did not find one.

File: lib/blacklight/access_controls/enforcement.rb [hunk 9]
```diff
@@ -1,22 +1,26 @@
...
       extend ActiveSupport::Concern
 
       included do
         extend Deprecation
-        attr_writer :current_ability
+        attr_writer :current_ability, :discovery_permissions
         deprecation_deprecate :current_ability=
 
         class_attribute :solr_access_filters_logic
+        alias_method :add_access_controls_to_solr_params, :apply_gated_discovery
```
Comment (cam156) [inline lib/blacklight/access_controls/enforcement.rb]: @atz Good use of alias here!

File: lib/blacklight/access_controls/enforcement.rb [hunk 10]
```diff
@@ -25,72 +29,61 @@ module Enforcement
...
-            user_access_filters << escape_filter(solr_field_for(type, 'user'), user.user_key)
-          end
+        return [] unless user && user.user_key.present?
+        permission_types.map do |type|
+          escape_filter(solr_field_for(type, 'user'), user.user_key)
         end
-        user_access_filters
       end
 
-      # Find the name of the solr field for this type of permission.
```
Comment (cam156) [inline lib/blacklight/access_controls/enforcement.rb]: @atz Not sure removing this comment is a great idea.  It does describe the overall method.
❤️  the break out of params below.

File: spec/unit/enforcement_spec.rb (hunk 2 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: That doesn't do what you think it does.  The class is still visible everywhere after the parser gets to it.  It is a false sense of scoping.

File: spec/unit/enforcement_spec.rb (hunk 3 above)
Comment (atz) [inline spec/unit/enforcement_spec.rb]: That's fine.  It's also clearly not important to any other piece of the stack.  In my opinion, it isn't important here.

File: lib/blacklight/access_controls/enforcement.rb (hunk 1 above)
Comment (atz) [inline lib/blacklight/access_controls/enforcement.rb]: As I mentioned, all production systems that I know about have a default user group (e.g. "public") and therefore will never experience this edge case.  The tests here and in the overall suite still connect to solr with real queries and complete fine.  If you want to strengthen the test suite for valid scenarios that you think are more likely to produce invalid queries, I would be happy to code to satisfy them.  But I don't need to speculate about query validity here.

File: lib/blacklight/access_controls/enforcement.rb (hunk 8 above)
Comment (atz) [inline lib/blacklight/access_controls/enforcement.rb]: This is below the level of PR review I regard as substantive.  The comment is *improved* by the addition.  It can be merged **as is** without detriment.  If you think it could be better still, you are probably correct!  But you are also welcome to submit further revisions, including against this branch.  IMO, the value represented by this PR is pretty well separate from the difference between the improved comment and a potentially better one.
